programs/marginfi/src/instructions/marginfi_account/borrow.rs

```diff
@@ -78,7 +85,38 @@ pub fn lending_account_borrow<'info>(
             .transpose()?
             .unwrap_or(amount);
 
-        bank_account.borrow(I80F48::from_num(amount_pre_fee))?;
+        let origination_fee_u64: u64;
+        if !origination_fee_rate.is_zero() {
+            let origination_fee: I80F48 = I80F48::from_num(amount_pre_fee)
+                .checked_mul(origination_fee_rate)
+                .ok_or_else(math_error!())?
+                .floor(); // round fees down in the user's favor
+            origination_fee_u64 = origination_fee.checked_to_num().ok_or_else(math_error!())?;
+
+            // Incurs a borrow that includes the origination fee (but withdraws just the amt)
+            bank_account.borrow(I80F48::from_num(amount_pre_fee) + origination_fee)?;
+
+            // The bank fee account gains the origination fee
+            bank_account.withdraw_spl_transfer(
+                origination_fee_u64,
+                bank_liquidity_vault.to_account_info(),
+                fee_vault.to_account_info(),
+                bank_liquidity_vault_authority.to_account_info(),
+                maybe_bank_mint.as_ref(),
+                token_program.to_account_info(),
+                bank_signer!(
+                    BankVaultType::Liquidity,
+                    bank_loader.key(),
+                    liquidity_vault_authority_bump
+                ),
+                ctx.remaining_accounts,
+            )?;
```

Comment:
Lets avoid doing a CPI call if not necessary, as they are super expensive. Regular fees are charged first in the accounting model, and are then manually withdrawn to the fee vault, this approach enables us to not worry about rounding and extra CPI costs on each hot action like borrowing.

Here we can simply add the collected fees to the `collected_group_fees_outstanding` in the bank. (https://github.com/mrgnlabs/marginfi-v2/blob/main/programs/marginfi/src/state/marginfi_group.rs#L323)

Then we can manually transfer the fee when we want.

Reply:
Great idea, done.